pkg/controller/actions/cacher/cacher.go

```diff
@@ -0,0 +1,68 @@
+package cacher
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/types"
+)
+
+// Renderer is the actual function received from upper layers
+// which generates resources based on the ReconciliationRequest.
+// Resource can be any type, so it returns `any` (or `error`)
+// The upper layers passing the function and can cast it to the expected type.
+type Renderer func(ctx context.Context, rr *types.ReconciliationRequest) (any, error)
+
+type CachingKeyFn func(rr *types.ReconciliationRequest) ([]byte, error)
+
+type Cacher struct {
+	cachingKeyFn    CachingKeyFn
+	cachingKey      []byte
+	cachedResources any
+}
+
+type CacherOpts func(*Cacher)
+
+func (s *Cacher) SetKeyFn(key CachingKeyFn) {
+	s.cachingKeyFn = key
+}
+
+// InvalidateCache invalidates both caching key and cached resources
+// what forces Render() to regenerate them.
+func (s *Cacher) InvalidateCache() {
+	s.cachingKey = nil
+	s.cachedResources = nil
+}
+
+func (s *Cacher) Render(ctx context.Context, r Renderer, rr *types.ReconciliationRequest) (any, bool, error) {
```

Comment:
nit: since the `Renderer` if a function, I would put it as the latest parameter to make it easier to define inline renderer functions, if needed

```go
r, a, err := c.Render(ctx, tt, func(ctx context.Context, rr *types.ReconciliationRequest) (any, error) {
   ...
})
```